File: helpdesk/src/main/java/com/leonidas/helpdesk/api/service/TicketService.java
package com.leonidas.HelpDesk.api.service;

import com.leonidas.HelpDesk.api.entity.ChangeStatus;
import com.leonidas.HelpDesk.api.entity.Ticket;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Component;

import java.util.Iterator;

@Component
public interface TicketService {

    Ticket createOrUpdate(Ticket ticket);

    Ticket findById(String id);

    void delete(String id);

    Page<Ticket> listTicket(int page, int count);

    ChangeStatus createChangeStatus(ChangeStatus changeStatus);

    Iterable<ChangeStatus> listChangeStatus(String ticketId);

    Page<Ticket> findByCurrentUser(int page, int count, String userId);

    Page<Ticket> findByParameters(int page, int count, String title, String status, String priority);

    Page<Ticket> findByParametersAndCurrentUser(int page, int count, String title, String status, String priority, String userId);

    Page<Ticket> findByNumber(int page, int count, Integer number);

    Iterable<Ticket> findAll();

    Page<Ticket> findByParameterAndAssignedUser(int page, int count, String title, String status, String priority, String assignedUser);
}

File: angular/helpdesk/src/app/services/ticket.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Ticket } from '../model/ticket.model';
import { HELP_DESK_API } from './helpdesk.api';
import { ResponseApi } from '../model/response-api';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class TicketService {

  constructor(private http: HttpClient) { }

  createOrUpdate(ticket: Ticket): Observable<ResponseApi>{
    if(ticket.id != null && ticket.id != ''){
      return this.http.put<ResponseApi>(`${HELP_DESK_API}/api/ticket`, ticket);
    } else {
      ticket.id = null;
      ticket.status = 'New';
      return this.http.post<ResponseApi>(`${HELP_DESK_API}/api/ticket`, ticket);
    }
  }

  findAll(page: number, count: number): Observable<ResponseApi> {
    return this.http.get<ResponseApi>(`${HELP_DESK_API}/api/ticket/${page}/${count}`);
  }

  findById(id: string): Observable<ResponseApi>{
    return this.http.get<ResponseApi>(`${HELP_DESK_API}/api/ticket/${id}`);
  }

  delete(id: string): Observable<ResponseApi>{
    return this.http.delete<ResponseApi>(`${HELP_DESK_API}/api/ticket/${id}`);
  }

  findByParams(page: number, count: number, assignedToMe: boolean, t: Ticket): Observable<ResponseApi>{
    t.number = t.number ?? 0;
    t.title = t.title == '' ? 'uninformed' : t.title;
    t.status = t.status == '' ? 'uninformed' : t.status;
    t.priority = t.priority == '' ? 'uninformed' : t.priority;

    return this.http.get<ResponseApi>(`${HELP_DESK_API}/api/ticket/${page}/${count}/${t.number}/${t.title}/${t.status}/${t.priority}/${assignedToMe}`);
  }

  changeStatus(status: string, ticket: Ticket){
    return this.http.put(`${HELP_DESK_API}/api/ticket/${ticket.id}/${status}`, ticket);
  }

  summary(){
    return this.http.get(`${HELP_DESK_API}/api/ticket/summary`);
  }

}
